src/compute_jacobian_ad.jl

```diff
@@ -0,0 +1,40 @@
+using ForwardDiff: Dual, jacobian, partials
+
+function compute_jacobian!(J::AbstractMatrix{<:Number},
+                f,
+                x::AbstractArray{<:Number};
+                color=1:length(x))
+
+    partials_array = Array{Float64}(undef, length(x), maximum(color))
+    for color_i in 1:maximum(color)
+        for i in 1:length(x)
+            if color[i]==color_i
+                partials_array[i,color_i] = 1
+            else
+                partials_array[i,color_i] = 0
+            end
+        end
+    end
+
+    p = Tuple.(eachrow(partials_array))
+    t = zeros(Dual{Nothing, Float64, maximum(color)},0)
```

Comment:
```suggestion
    t = zeros(Dual{typeof(f), Float64, maximum(color)},length(x))
```

It's usually better to pre-allocate. The first part here being `typeof(f)` will tag the Duals so they won't accidentally overlap with other differentiations.